pexv2/pd.go

```diff
@@ -22,6 +34,98 @@ type InputDescriptor struct {
 	Constraints Constraints `json:"constraints"`
 }
 
+type tokenizedField struct {
+	token string
+	path  string
+}
+
+// SelectCredentials selects vcJWTs based on the constraints defined in the input descriptor
+func (ind InputDescriptor) SelectCredentials(vcJWTs []string) ([]string, error) {
+	jsonSchema := JSONSchema{
+		Schema:     "http://json-schema.org/draft-07/schema#",
+		Type:       "object",
+		Properties: make(map[string]Filter, len(ind.Constraints.Fields)),
+		Required:   make([]string, 0, len(ind.Constraints.Fields)),
+	}
+
+	tokenizedFields := make([]tokenizedField, 0, len(ind.Constraints.Fields))
+	tokens := make([]string, len(ind.Constraints.Fields))
+
+	for i, field := range ind.Constraints.Fields {
+		token := strconv.FormatInt(time.Now().UnixNano(), 10)
+		tokens[i] = token
+
+		for _, path := range field.Path {
+			tf := tokenizedField{token: token, path: path}
+			tokenizedFields = append(tokenizedFields, tf)
+		}
+
+		if field.Filter != nil {
+			jsonSchema.AddProperty(token, *field.Filter)
+		}
+	}
+
+	sch, err := json.Marshal(jsonSchema)
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling schema: %w", err)
+	}
+
+	schema, err := jsonschema.CompileString(ind.ID, string(sch))
+	if err != nil {
+		return nil, fmt.Errorf("error compiling schema: %w", err)
+	}
+
+	matched := make([]string, 0, len(vcJWTs))
+
+	for _, vcJWT := range vcJWTs {
+		tokensFound := make(map[string]bool, len(tokenizedFields))
+
+		decoded, err := vc.Decode[vc.Claims](vcJWT)
+		if err != nil {
+			continue
+		}
+
+		var jwtPayload map[string]any
+		payload, err := base64.RawURLEncoding.DecodeString(decoded.JWT.Parts[1])
+		if err != nil {
+			continue
+		}
+
+		if err := json.Unmarshal(payload, &jwtPayload); err != nil {
+			continue
+		}
+
+		selectionCandidate := make(map[string]any)
+		for _, tf := range tokenizedFields {
+			if ok := tokensFound[tf.token]; ok {
+				continue
+			}
+
+			value, err := jsonpath.Get(tf.path, jwtPayload)
+			if err != nil {
+				continue
+			}
+
+			if value != nil {
+				selectionCandidate[tf.token] = value
+				tokensFound[tf.token] = true
+			}
+		}
+
+		if len(selectionCandidate) != len(tokens) {
```

Comment:
can you explain what this is doing?

Reply:
Yep. If the selection candidate doesn't have numTokens fields, that means it missed one of the required tokens (aka an expected field in the pd wasn't satisfied)